File: src/main/java/xyz/upperlevel/hgame/scenario/sequence/Sequence.java
package xyz.upperlevel.hgame.scenario.sequence;

import lombok.Getter;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

public class Sequence {
    private static List<Sequence> sequences = new ArrayList<>();

    private Map<Step, Trigger> incomingSteps = new HashMap<>();
    private List<Step> steps = new ArrayList<>();
    private List<Sequence> children = new ArrayList<>();

    private Sequence() {
    }

    /**
     * Appends an action to the current sequence.
     * To continue the sequence, the action is supposed to call the function {@link Sequence.Step#next(Trigger)}.
     */
    public Sequence append(Consumer<Step> action) {
        steps.add(new Step(steps.size()) {
            @Override
            public void start() {
                action.accept(this);
            }
        });
        return this;
    }

    public Sequence act(Runnable action) {
        append(step -> {
            action.run();
            step.next(Trigger.NONE);
        });
        return this;
    }

    public Sequence delay(long delay) {
        append(step -> step.next(Trigger.sleep(delay)));
        return this;
    }

    private void issueRepeat(Step root, Consumer<Step> action, long each) {
        append(step -> {
            action.accept(root);
            if (!root.isNextCalled()) {
                delay(each);
                issueRepeat(root, action, each);
                step.next(Trigger.NONE);
            } else {
                dismiss();
            }
        });
    }

    public Sequence repeat(Consumer<Step> action, long each) {
        append(root -> {
            // We create a sub-sequence that will handle the
            // repeating steps. If `next` is called within
            // the `action`, the first level sequence skips to the next step
            // and the inner sequence is dismissed.
            Sequence inner = new Sequence();
            inner.issueRepeat(root, action, each);
            children.add(inner);
            inner.play();
        });
        return this;
    }

    public Sequence repeat(BiConsumer<Step, Integer> action, long each, int times) {
        AtomicInteger wrapped = new AtomicInteger();
        repeat(step -> {
            action.accept(step, wrapped.get());
            if (wrapped.incrementAndGet() >= times) {
                step.next(Trigger.NONE);
            }
        }, each);
        return this;
    }

    public void update() {
        for (Entry<Step, Trigger> incomingStep : new HashMap<>(incomingSteps).entrySet()) {
            if (incomingStep.getValue().get()) {
                Step step = incomingStep.getKey();
                step.start();
                incomingSteps.remove(incomingStep.getKey());
            }
        }
        children.forEach(Sequence::update);
    }

    public void play() {
        if (steps.size() > 0) {
            steps.get(0).start();
        }
    }

    public void dismiss() {
        incomingSteps.clear();
        children.forEach(Sequence::dismiss);
        sequences.remove(this);
    }

    public abstract class Step {
        @Getter
        private boolean nextCalled;
        private int position;

        public Step(int position) {
            this.position = position;
        }

        public void next(Trigger trigger) {
            nextCalled = true;
            if (position >= steps.size() - 1) {
                return;
            }
            Step next = steps.get(position + 1);
            incomingSteps.put(next, trigger);
        }

        public abstract void start();
    }

    public static Sequence create() {
        Sequence sequence = new Sequence();
        sequences.add(sequence);
        return sequence;
    }

    public static void updateAll() {
        for (Sequence sequence : sequences) {
            sequence.update();
        }
    }

    public static void dismissAll() {
        for (Sequence sequence : sequences) {
            sequence.dismiss();
        }
    }
}

File: src/main/java/xyz/upperlevel/hgame/world/sequence/Sequence.kt
package xyz.upperlevel.hgame.world.sequence

import com.badlogic.gdx.Gdx
import org.apache.logging.log4j.LogManager
import java.util.*
import java.util.concurrent.atomic.AtomicInteger

class Sequence {
    private val incomingSteps = HashMap<Step, Trigger>()
    private val steps = ArrayList<Step>()
    private val children = ArrayList<Sequence>()

    /**
     * Appends an action to the current sequence.
     * To continue the sequence, the action is supposed to call the function [Sequence.Step.next].
     */
    fun append(action: (Step) -> Unit): Sequence {
        steps.add(object : Step(steps.size) {
            override fun start() {
                action(this)
            }
        })
        return this
    }

    fun act(action: () -> Unit): Sequence {
        append { step ->
            action()
            step.next(Triggers.NONE)
        }
        return this
    }

    fun pressKey(key: Int, action: () -> Unit): Sequence {
        append { step ->
            pressKey(key, action)
            action()
            step.next { Gdx.input.isKeyPressed(key) }
        }
        return this
    }

    fun pressKeyOnce(key: Int, action: () -> Unit): Sequence {
        append { step ->
            pressKeyOnce(key, action)
            action()
            step.next { Gdx.input.isKeyJustPressed(key) }
        }
        return this
    }

    fun delay(delay: Long): Sequence {
        append { step -> step.next(Triggers.sleep(delay)) }
        return this
    }

    private fun issueRepeat(root: Step, action: (Step) -> Unit, each: Long) {
        append { step ->
            action(root)
            if (!root.isNextCalled) {
                delay(each)
                issueRepeat(root, action, each)
                step.next(Triggers.NONE)
            } else {
                dismiss()
            }
        }
    }

    fun repeat(action: (Step) -> Unit, each: Long): Sequence {
        append { root ->
            // We create a sub-sequence that will handle the
            // repeating steps. If `next` is called within
            // the `action`, the first level sequence skips to the next step
            // and the inner sequence is dismissed.
            val inner = Sequence()
            inner.issueRepeat(root, action, each)
            children.add(inner)
            inner.play()
        }
        return this
    }

    fun repeat(action: (Step, Int) -> Unit, each: Long, times: Int): Sequence {
        val wrapped = AtomicInteger()
        repeat({ step ->
            action(step, wrapped.get())
            // If 'times' is negative, it'll repeat the loop infinitely,
            // The counter will increment starting from 0.
            if (wrapped.incrementAndGet() >= times && times >= 0) {
                step.next(Triggers.NONE)
            }
        }, each)
        return this
    }

    fun update() {
        for ((step, value) in HashMap(incomingSteps)) {
            if (value()) {
                step.start()
                incomingSteps.remove(step)
            }
        }
        children.forEach { it.update() }
    }

    fun play(): Sequence {
        if (steps.size > 0) {
            steps[0].start()
        }
        return this
    }

    fun dismiss() {
        children.forEach { it.dismiss() }
        sequences.remove(this)
    }

    abstract inner class Step(private val position: Int) {
        var isNextCalled: Boolean = false
            private set

        fun next(trigger: Trigger) {
            isNextCalled = true
            if (position >= steps.size - 1) {
                return
            }
            val next = steps[position + 1]
            incomingSteps[next] = trigger
        }

        abstract fun start()
    }

    companion object {
        private val sequences = ArrayList<Sequence>()
        private val logger = LogManager.getLogger()

        fun create(): Sequence {
            val sequence = Sequence()
            sequences.add(sequence)
            return sequence
        }

        fun updateAll() {
            for (sequence in ArrayList(sequences)) {
                sequence.update()
            }
        }

        fun dismissAll() {
            for (sequence in ArrayList(sequences)) {
                sequence.dismiss()
            }
        }
    }
}
